src/main/java/org/opensearch/security/privileges/ActionPrivileges.java

```diff
@@ -0,0 +1,1146 @@
+                )
+                .evaluationExceptions(exceptions);
+        }
+
+        /**
+         * Returns PrivilegesEvaluatorResponse.ok() if the user identified in the context object has privileges for all
+         * indices (using *) for the given actions. Returns null otherwise. Then, further checks must be done to check
+         * the user's privileges.
+         */
+        PrivilegesEvaluatorResponse providesWildcardPrivilege(PrivilegesEvaluationContext context, Set<String> actions) {
+            ImmutableSet<String> effectiveRoles = context.getMappedRoles();
+
+            for (String action : actions) {
+                ImmutableCompactSubSet<String> rolesWithWildcardIndexPrivileges = this.actionToRolesWithWildcardIndexPrivileges.get(action);
+
+                if (rolesWithWildcardIndexPrivileges == null || !rolesWithWildcardIndexPrivileges.containsAny(effectiveRoles)) {
+                    return null;
+                }
+            }
+
+            return PrivilegesEvaluatorResponse.ok();
+        }
+
+        /**
+         * Checks whether this instance provides explicit privileges for the combination of the provided action,
+         * the provided indices and the provided roles.
+         * <p>
+         * Explicit means here that the privilege is not granted via a "*" action privilege wildcard. Other patterns
+         * are possible. See also: https://github.com/opensearch-project/security/pull/2411 and https://github.com/opensearch-project/security/issues/3038
+         */
+        PrivilegesEvaluatorResponse providesExplicitPrivilege(
+            PrivilegesEvaluationContext context,
+            Set<String> actions,
+            IndexResolverReplacer.Resolved resolvedIndices,
+            CheckTable<String, String> checkTable,
+            Map<String, IndexAbstraction> indexMetadata
+        ) {
+            List<PrivilegesEvaluationException> exceptions = new ArrayList<>();
+
+            if (!CollectionUtils.containsAny(actions, this.explicitlyRequiredIndexActions)) {
+                return PrivilegesEvaluatorResponse.insufficient(CheckTable.create(ImmutableSet.of("_"), actions), context);
+            }
+
+            for (String role : context.getMappedRoles()) {
+                ImmutableMap<String, IndexPattern> actionToIndexPattern = this.rolesToExplicitActionToIndexPattern.get(role);
+
+                if (actionToIndexPattern != null) {
+                    for (String action : actions) {
+                        IndexPattern indexPattern = actionToIndexPattern.get(action);
+
+                        if (indexPattern != null) {
+                            for (String index : checkTable.iterateUncheckedRows(action)) {
+                                try {
+                                    if (indexPattern.matches(index, context, indexMetadata) && checkTable.check(index, action)) {
+                                        return PrivilegesEvaluatorResponse.ok();
```

Comment:
For my own understanding, if `checkTable.check(index, action)` returns true that means all cells in the table have been checked and the table is complete?

Reply:
Exactly. I should have added that as a comment - will do.